### ISSUE TITLE
Surfacing global variables

### ISSUE THREAD

**Issue by FayCross** [thexerteproject/xerteonlinetoolkits#970]
I've just committed a change allowing variables in Bootstrap & XOT templates to be surfaced in projects using format {varName}. This will work for things like {x_lastUpdatedLocale} but there are probably lots of other useful variables we could create in order to be surfaced in this way.

- No. project plays (other tracking data?)

- Creator / editor names

**Comment by ronm123** [comment]
@FayCross sounds good. However I've pulled down the changes and added {x_lastUpdatedLocale} to a page in an XOT project but it's just displaying that as {x_lastUpdatedLocale}
I've cleared offline files and refreshed - what am I missing?

As for other variables I think perhaps original creation date in addition to the last updated date might be useful.

**Comment by FayCross** [comment]
There's an opt property at LO level that turns this on. We could have it on for all projects by default if you'd prefer - I just thought it might be best to avoid doing the checking through all text fields for markup if if wasn't needed.

**Comment by ronm123** [comment]
@FayCross ah sorry didn't spot that. Didn't look for it but good idea to include it as an option. Yes probably best off by default for the reason you describe. At some point we'll need to list what they are somewhere too.

**Comment by FayCross** [comment]
Using {lastUpdated} {dateCreated} {numViews} will now load this info in projects. Variables using x_ format still exist but I've also added these simple var names following @JohnSmith-LT comments yesterday. Do we want to restrict surfacing of variables other than these? Is there any harm in allowing surfacing of any global vars?

**Comment by JohnSmith-LT** [comment]
Hi @FayCross , don't think there is any harm in it, I just think that we've backed ourself into a corner on several occasions where there are things now that we can't change for fear of breaking future projects... My thinking is that at some point we will probably want/need to restructure xenith and make it modular and then a lot of the variables will not be accessible because they will be behind a closure/function... hence I was thinking if we made an object such as XVARS (probably not the best name...) and had XVARS.lastUpdated, etc and these were documented use these but use the others "at your peril!!" then it would be relatively easy for us just to populate that either from within a xenith "module" or from the php play/preview code... let them use anything else they can find, like x_pageInfo.pageID but to be aware that at some point it might not be available... otherwise we can never restructure because @torinfo and @ronm123 will have 100s of projects using this ;-)

**Comment by FayCross** [comment]
No that makes sense re. a variables object.
I knew someone would ask about getting them to update... there's nothing to do this at the moment but it does stick a span around the text that's replaced the {...} markup with a 'globalVars' class. If we gave the span some info about specifically what it's replaced then we could get them to update on subsequent page views. The variables code already does this so it wouldn't be hard to add

**Comment by FayCross** [comment]
Changed so it's now globalVars.lastUpdated, globalVars.dateCreated & globalVars.numViews Not sure globalVars is best name but can be changed before next release if needed

**Comment by ronm123** [comment]
Hi all
I noticed an issue with this as a result of new functionality we have and debated whether or not to just start a new issue or to re-open this one even though it's new because of new options. But anyway have decided to add it here:

I wanted to add some of the global variables to a project and reminded myself how we do it in the footer of the release notes. It seems in that resource we have some script to replace the last updated value in the footer. I'm not sure if we added/used that before the surface global variables functionally was added (and it works as it is) however using additional script isn't something many others are likely to want or be able to do so I tested how we say it should work in the release notes for v3.10 e.g. you might add the following to a text block in either Bootstrap or XOT:

`First created: {globalVars.dateCreated} Last updated: {globalVars.lastUpdated} View count: {globalVars.numViews}`

If added to a text block within a page in Bootstrap or XOT that works and the correct values get inserted.
However if you add the above to the custom footer option in Bootstrap, where you're arguably most likely to want it, then the values don't get inserted. I think that's because the current function in application.js doesn't attempt to replace the values in the custom footer. Not sure if we should extend that existing function so it does work or add a new function just for the footer.

I also tested adding the same line above to a few different places in an XOT project. Again it does work when added to normal text boxes that display on a page by default but we now have more functionality and so for example it doesn't work when added to the following:
project introduction
page introduction
resources page
and these are all perhaps places where users might want to display at least last update etc.
Again I think this is probably because the current function in xenith doesn't look in all these places.

In short two desirable improvements:

- [ ] Extend the function in Bootstrap application.js so that these global vars can work in the footer without needing custom code.
- [ ] Extend the function in XOT xenith.js so that the global vars can be used in the newer options as listed above.
